File: VB/RichEdit_InsertChart/ChartImage.vb
Imports DevExpress.XtraRichEdit.API.Native
Imports System.Xml
Imports System.Text
Imports DevExpress.XtraRichEdit.Utils
Imports System.Drawing.Imaging
Imports DevExpress.XtraCharts
Imports System.IO
Imports DevExpress.XtraRichEdit.Model
Imports DevExpress.XtraRichEdit.Fields
Imports System.Linq
Imports System.Collections.Generic
Imports System
Imports DevExpress.Utils

Namespace RichEdit_InsertChart

	Public Class ChartImage


		Private _Input As String
		Private _ViewType As DevExpress.XtraCharts.ViewType = DevExpress.XtraCharts.ViewType.Pie
		Private _Width? As Integer = Nothing
		Private _Height? As Integer = Nothing
		Private _ShowLegend As DefaultBoolean = DefaultBoolean.False
		Private _Data As New List(Of KeyValuePair(Of String, Double))()

		Public ReadOnly Property ViewType() As DevExpress.XtraCharts.ViewType
			Get
				Return Me._ViewType
			End Get
		End Property
		Public ReadOnly Property Width() As Integer?
			Get
				Return Me._Width
			End Get
		End Property
		Public ReadOnly Property Height() As Integer?
			Get
				Return Me._Height
			End Get
		End Property
		Public ReadOnly Property ShowLegend() As DefaultBoolean
			Get
				Return Me._ShowLegend
			End Get
		End Property
		Public ReadOnly Property Data() As List(Of KeyValuePair(Of String, Double))
			Get
				Return _Data
			End Get
		End Property


		Public Sub New(ByVal input As String)
			Me._Input = input
		End Sub

		Public Sub Initialize()
			Using stringReader As New StringReader(_Input)
				Using reader As New XmlTextReader(stringReader)
					reader.DtdProcessing = DtdProcessing.Ignore
					Do While reader.Read()
						If reader.IsStartElement() Then
							Select Case reader.Name
								Case "type"
									GetChartType(reader.ReadString())
								Case "height"
									_Height = New Integer?(Integer.Parse(reader.ReadString()))
								Case "width"
									_Width = New Integer?(Integer.Parse(reader.ReadString()))
								Case "legend"
									_ShowLegend = If((reader.ReadString().ToLower() = "true"), DefaultBoolean.True, DefaultBoolean.False)
								Case "data"
									GetData(reader.ReadString())
							End Select
						End If
					Loop
				End Using
			End Using
		End Sub

		Private Sub GetData(ByVal p As String)
			For Each t As String In p.Split(","c)
				Dim value As Double

				Dim pair() As String = t.Split("|"c)

				If pair.Length = 0 Then
					Continue For
				End If

				If pair.Length = 1 Then
					If Not Double.TryParse(pair(0), value) Then
						_Data.Add(New KeyValuePair(Of String, Double)(pair(0), 0))
					Else
						_Data.Add(New KeyValuePair(Of String, Double)(String.Empty, value))
					End If
				Else
					If pair.Length = 2 Then
						If Not Double.TryParse(pair(1), value) Then
							_Data.Add(New KeyValuePair(Of String, Double)(pair(0), 0))
						Else
							_Data.Add(New KeyValuePair(Of String, Double)(pair(0), value))
						End If
					End If
				End If
			Next t
		End Sub

		Private Sub GetChartType(ByVal p As String)
			Select Case p.ToLower()
				Case "bar"
					_ViewType = DevExpress.XtraCharts.ViewType.Bar
				Case "line"
					_ViewType = DevExpress.XtraCharts.ViewType.Line
				Case "pie3d"
					_ViewType = DevExpress.XtraCharts.ViewType.Pie3D
				Case Else
					_ViewType = DevExpress.XtraCharts.ViewType.Pie
			End Select
		End Sub


		Public Function CreateChart() As Stream

			Using chart As New ChartControl()
				If _Width.HasValue Then
					chart.Width = _Width.Value
				End If
				If _Height.HasValue Then
					chart.Height = _Height.Value
				End If

				Dim undefined As Integer = 1
				Dim stream As New MemoryStream()
				Try
					Dim series As New Series("Chart", _ViewType)
					Try
						If TypeOf series.Label Is DevExpress.XtraCharts.PieSeriesLabel Then
							CType(series.Label, DevExpress.XtraCharts.PieSeriesLabel).Position = PieSeriesLabelPosition.Inside
						End If

						If _ViewType = ViewType.Pie Then
							series.Label.TextPattern = "{A}: {VP:P0}"
						End If

						If _Data Is Nothing OrElse _Data.Count = 0 Then
							series.Points.Add(New SeriesPoint("Undefined", New Double() {1}))
							series.Label.TextPattern = "{S}"

						Else
							series.Label.TextPattern = "{A}: {V:F1}"

							For i As Integer = 0 To _Data.Count - 1

								Dim argument As String = _Data(i).Key.Trim()

								If String.IsNullOrEmpty(argument) Then
									argument = "Undefined " & undefined
									undefined += 1
								End If

								series.Points.Add(New SeriesPoint(argument, New Double() {_Data(i).Value}))
							Next i
						End If

						chart.Legend.Visibility = _ShowLegend
						chart.BorderOptions.Visibility = DevExpress.Utils.DefaultBoolean.False

						chart.Series.AddRange(New Series() {series})
						series = Nothing

						Dim diagram As XYDiagram = TryCast(chart.Diagram, XYDiagram)
						If diagram IsNot Nothing AndAlso diagram.AxisX IsNot Nothing Then
							diagram.AxisX.Label.ResolveOverlappingOptions.AllowStagger = True
						End If


						chart.ExportToImage(stream, ImageFormat.Png)
						stream.Position = 0

						Return stream

					Catch
						If series IsNot Nothing Then
							series.Dispose()
						End If
						Throw
					End Try

				Catch
					If stream IsNot Nothing Then
						stream.Dispose()
					End If
					Throw
				End Try
			End Using
		End Function

		Public Function CreateImage() As DocumentImageSource
			Return DocumentImageSource.FromStream(CreateChart())
		End Function

	End Class
End Namespace

File: CS/RichEdit_InsertChart/ChartImage.cs
namespace RichEdit_InsertChart{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using DevExpress.XtraRichEdit.Fields;
    using DevExpress.XtraRichEdit.Model;
    using System.IO;
    using DevExpress.XtraCharts;
    using System.Drawing.Imaging;
    using DevExpress.XtraRichEdit.Utils;
    using System.Text;
    using System.Xml;
    using DevExpress.XtraRichEdit.API.Native;

    public class ChartImage {


        string _Input;
        DevExpress.XtraCharts.ViewType _ViewType = DevExpress.XtraCharts.ViewType.Pie;
        Nullable<int> _Width = null;
        Nullable<int> _Height = null;
        DevExpress.Utils.DefaultBoolean _ShowLegend = DevExpress.Utils.DefaultBoolean.False;
        List<KeyValuePair<string, double>> _Data = new List<KeyValuePair<string, double>>();

        public DevExpress.XtraCharts.ViewType ViewType { get { return this._ViewType; } }
        public Nullable<int> Width { get { return this._Width; } }
        public Nullable<int> Height { get { return this._Height; } }
        public DevExpress.Utils.DefaultBoolean ShowLegend { get { return this._ShowLegend; } }
        public List<KeyValuePair<string, double>> Data { get { return _Data; } }      
        
        
        public ChartImage(string input)
        {
            this._Input = input;
        }

        public void Initialize()
        {
            using (StringReader stringReader = new StringReader(_Input))
            using (XmlTextReader reader = new XmlTextReader(stringReader)) {
                reader.DtdProcessing = DtdProcessing.Ignore;
                while (reader.Read()) {
                    if (reader.IsStartElement()) {
                        switch (reader.Name) {
                            case "type":
                                GetChartType(reader.ReadString());
                                break;
                            case "height":
                                _Height = new int?(int.Parse(reader.ReadString()));
                                break;
                            case "width":
                                _Width = new int?(int.Parse(reader.ReadString()));
                                break;
                            case "legend":
                                _ShowLegend = (reader.ReadString().ToLower() == "true") ? DevExpress.Utils.DefaultBoolean.True : DevExpress.Utils.DefaultBoolean.False;
                                break;
                            case "data":
                                GetData(reader.ReadString());
                                break;
                        }
                    }
                }
            }
        }

        private void GetData(string p)
        {
            foreach (string t in p.Split(',')) {
                double value;

                string[] pair = t.Split('|');

                if (pair.Length == 0) {
                    continue;
                }

                if (pair.Length == 1) {
                    if (!double.TryParse(pair[0], out value)) {
                        _Data.Add(new KeyValuePair<string, double>(pair[0], 0));
                    }
                    else {
                        _Data.Add(new KeyValuePair<string, double>(String.Empty, value));
                    }
                }
                else
                    if (pair.Length == 2) {
                        if (!double.TryParse(pair[1], out value)) {
                            _Data.Add(new KeyValuePair<string, double>(pair[0], 0));
                        }
                        else {
                            _Data.Add(new KeyValuePair<string, double>(pair[0], value));
                        }
                    }
            }
        }

        private void GetChartType(string p)
        {
            switch (p.ToLower()) {
                case "bar":
                    _ViewType = DevExpress.XtraCharts.ViewType.Bar;
                    break;
                case "line":
                    _ViewType = DevExpress.XtraCharts.ViewType.Line;
                    break;
                case "pie3d":
                    _ViewType = DevExpress.XtraCharts.ViewType.Pie3D;
                    break;
                case "pie":
                default:
                    _ViewType = DevExpress.XtraCharts.ViewType.Pie;
                    break;
            }
        }


        public Stream CreateChart() {

            using (ChartControl chart = new ChartControl()) {
                if (_Width.HasValue) {
                    chart.Width = _Width.Value;
                }
                if (_Height.HasValue) {
                    chart.Height = _Height.Value;
                }

                int undefined = 1;
                MemoryStream stream = new MemoryStream();
                try {
                    Series series = new Series("Chart", _ViewType);
                    try {
                        if (series.Label is DevExpress.XtraCharts.PieSeriesLabel) {
                            ((DevExpress.XtraCharts.PieSeriesLabel)series.Label).Position = PieSeriesLabelPosition.Inside;
                        }

                        if (_ViewType == ViewType.Pie) {
                            series.Label.TextPattern = "{A}: {VP:P0}";
                        }

                        if (_Data == null || _Data.Count == 0) {
                            series.Points.Add(new SeriesPoint("Undefined", new double[] { 1 }));
                            series.Label.TextPattern = "{S}";

                        } else {
                            series.Label.TextPattern = "{A}: {V:F1}";

                            for (int i = 0; i < _Data.Count; i++) {

                                string argument = _Data[i].Key.Trim();

                                if (String.IsNullOrEmpty(argument)) {
                                    argument = "Undefined " + undefined;
                                    undefined++;
                                }

                                series.Points.Add(new SeriesPoint(argument, new double[] {
                                    _Data[i].Value }));
                            }
                        }

                        chart.Legend.Visibility = _ShowLegend;
                        chart.BorderOptions.Visibility = DevExpress.Utils.DefaultBoolean.False;

                        chart.Series.AddRange(new Series[] { series });
                        series = null;

                        XYDiagram diagram = chart.Diagram as XYDiagram;
                        if (diagram != null && diagram.AxisX != null)
                            diagram.AxisX.Label.ResolveOverlappingOptions.AllowStagger = true; ;

                        chart.ExportToImage(stream, ImageFormat.Png);
                        stream.Position = 0;

                        return stream;

                    } catch {
                        if (series != null) {
                            series.Dispose();
                        }
                        throw;
                    }

                } catch {
                    if (stream != null) {
                        stream.Dispose();
                    }
                    throw;
                }
            }
        }

        public DocumentImageSource CreateImage()
        {
            return DocumentImageSource.FromStream(CreateChart());
        }

    }
}
